### PR TITLE
Adds rock.reduce max support

### PR DESCRIPTION
**Pull request by manupak**
This commit adds support rock.reduce max operation
with tests.

depends on #972.

closes: https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/775

### CODE REVIEW

**Comment by manupak** [comment]
Putting to draft until the dependent PR is merged.

File: mlir/lib/Dialect/Rock/Transforms/SugarToLoops.cpp [hunk 1]
```diff
@@ -893,6 +895,77 @@ struct BufferLoadRewritePattern : public OpRewritePattern<BufferLoadOp> {
...
+    Value sign =
+        b.create<arith::AndIOp>(loc, signbitConstantOp, stValIntCasted);
+    auto isPos = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                         zeroConstantOp, sign);
+
+    scf::IfOp ifb = b.create<scf::IfOp>(loc, isPos, /*withElseRegion=*/true);
+    {
+      OpBuilder thenb = ifb.getThenBodyBuilder(b.getListener());
+      // If the current value is positive doing a signed max integer comparison
+      // will be equivalent to a floating point max comparison.
```
Comment (sjw36) [inline mlir/lib/Dialect/Rock/Transforms/SugarToLoops.cpp:930]: Ah, very clever!

**Review by sjw36 — APPROVED** [review]
Looks great!

**Comment by sjw36** [comment]
@manupak how will we know to generate an init kernel with max negative value?

**Review by krzysz00 — COMMENTED** [review]
To echo @sjw36 , is changing `zero_init_kernel` to `constant_init_kernel` or the like a future PR?

**Comment by manupak** [comment]
Short term : 

We just need a way to indicate to clients or our testing infra that needs to be what value it needs to be memset/prefilled. We could do this today with something like : 

https://github.com/ROCmSoftwarePlatform/rocMLIR/blob/8c60ac9d8c3b2bed16a8fcd8243a1e8e96e1c74e/mlir/lib/Conversion/TosaToRock/TosaToRock.cpp#L468-L469

We just need to put negative inf (which is a value in IEEE 754) here.
This clone harness population logic will understand this and insert linalg.fill before the launch. 

Long term : 

If we want to generate memset kernels, we might need @krzysz00 what proposing among the whole shebang in the GH discussion : https://github.com/ROCmSoftwarePlatform/rocMLIR/discussions/940

**Comment by krzysz00** [comment]
When'd "prefillAttr" show up? I don't remember that landing

**Comment by manupak** [comment]
Well it was done ages ago.. was blocked on many things in the middle... And then we landed recently.

I think we discussed this as stop gap to get things going..?

**Comment by krzysz00** [comment]
I don't think we did? I remember us discussing it as an alternative to the memset kernel, but not actually decoding to go for it? 

I'm not opposed, mind - it's probably fine for MIGraphX to do the prefill thing as a short-term measure.

**Comment by manupak** [comment]
Well we certainly discussed in the team meeting of doing that i.e. putting the milestone. I remember, you asking the same question are we going for this route? Likewise the reasoning as a short term solution.

Anyways, Im not opposed to the long term solution either (well we co-designed it). Happy to remove this when we have that.

File: mlir/test/lit.site.cfg.py.in [hunk 2]
```diff
@@ -81,9 +113,8 @@ if config.rocm_path:
             agents = set(x.decode("utf-8") for x in q.stdout.split() if x != b"gfx000")
         config.arch = ','.join(agents)
         for x in agents:
-            if "gfx908" in x or "gfx90a" in x:
-                config.arch_support_mfma = True
-            # Check other features here
+            if not config.features:
+                set_arch_features(x)
```
Comment (manupak) [inline mlir/test/lit.site.cfg.py.in:117]: @zhanglx13 I did this to cut down on test code duplication to obtain features. Would you be able to take a look to see if thats alright ?

**Review by manupak — COMMENTED** [review]
OK taking inspiration from comments of @krzysz00 elsewhere, attempted a test parameterization here (also could be used anywhere now)

File: mlir/test/Dialect/Rock/integration/reduce/reduce_max/rock-reduce-max-case4.mlir [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// This test is checking for larger reductions with larger block and grid sizes
+
+// RUN: sed -e 's/#arch/%arch/g; s/#features/%features/g' %s | rocmlir-gen -ph -print-results -fut test_reduce -verifier clone - | rocmlir-driver -host-pipeline xmodel -kernel-pipeline full | xmir-runner --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_c_runner_utils%shlibext --entry-point-result=void | FileCheck %s --check-prefix=CLONE
```
Comment (krzysz00) [inline mlir/test/Dialect/Rock/integration/reduce/reduce_max/rock-reduce-max-case4.mlir]: I'd go for a rather stronger `sed` token, something like `##ARCH_SUBST##`

File: mlir/test/lit.site.cfg.py.in [hunk 4]
```diff
@@ -63,11 +63,43 @@ except KeyError:
     key, = e.args
     lit_config.fatal("unable to find %r parameter, use '--param=%s=VALUE'" % (key,key))
 
+# Helper function to decode arch to its features
+# Keep this in sync with mlir/lib/Dialect/Rock/Generator/AmdArchDb.cpp:mlir::rock::lookupArchInfo
+def set_arch_features(arch: str):
```
Comment (krzysz00) [inline mlir/test/lit.site.cfg.py.in:68]: You'll want comments on `AmdArchDb.cpp` as well, I think.

Comment (manupak) [reply mlir/test/lit.site.cfg.py.in:68]: I already have those!

**Comment by krzysz00** [comment]
Re the tests, would it perhaps make more sense to write them in Tosa and send them through the lowering flow instead of trying to substitute in features?

File: mlir/test/lit.site.cfg.py.in (hunk 2 above)
Comment (zhanglx13) [reply mlir/test/lit.site.cfg.py.in:117]: I think it looks very good and now we have a place to play with features.

**Comment by manupak** [comment]
I prefer to do the TOSA hook up seperately and this was done before we had a way to test them using tosa. 
I dont see a drawback in having these test or access to features in the meantime.

**Comment by krzysz00** [comment]
Yeah, I don't see any huge problems here

File: mlir/test/Dialect/Rock/integration/reduce/reduce_max/rock-reduce-max-case4.mlir (hunk 3 above)
Comment (manupak) [reply mlir/test/Dialect/Rock/integration/reduce/reduce_max/rock-reduce-max-case4.mlir]: done!

**Pull request merged**